File: python/cote/LTCD/조합.py
from typing import List
import itertools
import random

TEST_CASES = 10
N = 10
K = 5

def solve(nums: List[int], k: int) -> List[List[int]] :
	ret = []
	
	def dfs(elements: List[int], start: int, k: int) :
		if k == 0 :
			ret.append(elements[:])
			return

		for i, v in enumerate(nums) :
			if i < start : continue
			elements.append(v)
			dfs(elements, i + 1, k - 1)
			elements.pop()

		return

	dfs([], 0, k)
	return ret

def solve2(nums: List[int], k: int) -> List[List[int]] :
	return list(itertools.combinations(nums, k))

if __name__ == "__main__" :
	random.seed(43)

	for _ in range(TEST_CASES) :
		n = random.randrange(N) + 1
		k = random.randrange(min(n, K)) +1
		nums = random.sample(range(N), n)
		print(solve(nums, k), solve2(nums, k))

File: javascript/cote/LTCD/조합.mjs

const TEST_CASES = 10
const N = 10
const K = 5

function solve(nums, k) {
	let [ret, stack] = [[],[]]

	function dfs(start) {
		if (stack.length === k) {
			ret.push([...stack])
			return
		}

		for (let i = start; i < nums.length; i++) {
			stack.push(nums[i])
			dfs(i + 1)
			stack.pop()
		}
		return 
	}

	dfs(0)

	return ret
}

function main() {
	for (let it = 0; it < TEST_CASES; it++) {
		const k = Math.floor(Math.random() * (K - 1)) + 1
		const n = Math.floor(Math.random() * (N - k)) + k
		const nums = new Array(n).fill(0)
			.map((_, i) => i)

		console.log(solve(nums, k))
	}
}

main()